ports/stm32f4/common-hal/busio/I2C.c

```diff
@@ -32,23 +32,62 @@
 
 #include "shared-bindings/microcontroller/__init__.h"
 #include "supervisor/shared/translate.h"
+#include "stm32f4/periph.h"
 
-I2C_HandleTypeDef hi2c2;
+I2C_HandleTypeDef hi2c;
 
 void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
         const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout) {
     //TODO: Rework this entire section to use LL so we can properly assign pins
-    //      This will also be bundled with MSP removal
-    hi2c2.Instance = I2C1;
-    hi2c2.Init.ClockSpeed = 100000;
-    hi2c2.Init.DutyCycle = I2C_DUTYCYCLE_2;
-    hi2c2.Init.OwnAddress1 = 0;
-    hi2c2.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
-    hi2c2.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
-    hi2c2.Init.OwnAddress2 = 0;
-    hi2c2.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
-    hi2c2.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
-    if(HAL_I2C_Init(&hi2c2) != HAL_OK) {
+    //      This will also be bundled with MSP removal  
+    I2C_TypeDef * I2Cx = NULL;
+    I2C_TypeDef * I2Cx_check = NULL;
+
+    for(uint i=0; i<(sizeof(mcu_i2c_list)/sizeof(*mcu_i2c_list));i++) {
+        if (mcu_i2c_list[i]->sda_pin == sda) {
+            switch (mcu_i2c_list[i]->i2c_index) {
+                case 1: I2Cx = I2C1;
+                        break;
+                #ifdef I2C2
+                case 2: I2Cx = I2C2;
+                        break;
+                #endif
+                #ifdef I2C3
+                case 3: I2Cx = I2C3;
+                        break;
+                #endif
+            }
```

Comment:
Wait to map the i2c index to a pointer until after you know they match. That way you only need to do it once. Furthermore, use a table to do the mapping instead of a switch. If you need to validate the index, then do it before the pointer mapping.

Reply:
Something like this? 
```
#ifdef I2C3
    I2C_TypeDef * I2Cbanks[3] = {I2C1, I2C2, I2C3};
    #elif I2C1
    I2C_TypeDef * I2Cbanks[1] = {I2C1};
    #endif

    for(uint i=0; i<(sizeof(mcu_i2c_list)/sizeof(*mcu_i2c_list));i++) {
        if (mcu_i2c_list[i]->sda_pin == sda) {
            for(uint j=0; j<(sizeof(mcu_i2c_list)/sizeof(*mcu_i2c_list));j++) {
                if (mcu_i2c_list[i]->scl_pin == scl) {
                    if(mcu_i2c_list[i]->i2c_index == mcu_i2c_list[i]->i2c_index) {
                        I2Cx = I2Cbanks[mcu_i2c_list[i]->i2c_index-1];
                    }
                }
            }
        }
    }
```

Reply:
Actually that lookup table needs to be way, way uglier. There's no indication other than manually counting for how many indexes of a peripheral there are, and they can skip units. I don't think a lookup is the right option here.

Reply:
I could do it with a set size lookup like this, I guess
```
I2C_TypeDef * I2Cbanks[3] = 
    {
    #ifdef I2C1
        I2C1, 
    #else 
        NULL,
    #endif
    #ifdef I2C2
        I2C2,
    #else
        NULL,
    #endif
    #ifdef I2C3
        I2C3
    #else
        NULL
    #endif
    };
```